### PR TITLE
API routes with tests

### PR DESCRIPTION
**Pull request by tbrinkman3**

### CODE REVIEW

File: testHelpers/hasAllKeys.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+const hasAllKeys = (obj, array) => {
+  let objKeys = Object.keys(obj);
+
+  for ( let i = 0; i < objKeys.length; i++) {
+    if (objKeys[i] !== array[i]) {
+      return false
+    }
```
Comment (jenniferjinyoungyang) [inline testHelpers/hasAllKeys.js:7]: Would it be possible if Object.keys(obj) has the same keys as array but in different order? If so, would you want to return false?

File: servers/mongo/server-mongo.js [hunk 2]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
+  let q_id = Number(params.question_id)
+  let page = Number(query.page)
+  let count = Number(query.count)
+
+  db.getAnswers(q_id, page, count).then(data => {
+    res.status(200).send(data)
+  })
+  .catch(err => {
+    console.log('error getting answers', err)
+    res.send(400)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:43]: res.sendStatus(400)?

File: servers/mongo/server-mongo.js [hunk 3]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
+
+app.get('/qa/questions/:question_id/answers', (req, res) => {
+  let query = req.query;
+  let params = req.params
+
+  let q_id = Number(params.question_id)
+  let page = Number(query.page)
+  let count = Number(query.count)
+
+  db.getAnswers(q_id, page, count).then(data => {
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:38]: put ".then(data => {" next line?

File: servers/mongo/server-mongo.js [hunk 4]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
+  let name = query.name;
+  let email = query.email;
+  let p_id = query.product_id;
+
+  db.postQuestion(p_id, name, email, body)
+  .then(data => {
+    res.status(201).send(data)
+  })
+  .catch(err => {
+    console.log('error posting questions', err)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:60]: Maybe add res.sendStatus?

File: servers/mongo/server-mongo.js [hunk 5]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
+        return (l !== '[' && l !== ']')
+      }).join('').split(',')
+
+      console.log('photos', photos)
+
+      if (!photos) {
+        throw new Error('Incompatible data type for photos')
+      }
+    } catch (e) {
+      console.error(e)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:86]: If photos data type is incompatible, do you intend not to proceed with the post request? Should you end request here with something like res.sendStatus?

File: servers/mongo/server-mongo.js [hunk 6]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
+    }
+  }
+
+  db.postAnswer(q_id, body, name, email, photos)
+  .then(data =>{
+    res.status(201).send(data)
+  })
+  .catch(err => {
+    console.log('error posting answer', err)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:95]: Maybe add res.sendStatus?

File: servers/mongo/server-mongo.js [hunk 7]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
-app.post('/qa/questions')
+  let q_id = Number(params.question_id);
 
-app.post('/qa/questions/[questionid]/answers')
+  db.markQuestionHelpful(q_id)
+  .then(() => {
+    res.status(204).send()
+  })
+  .catch(err => {
+    console.log('error marking question helpful', err)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:109]: Maybe add res.sendStatus?

File: servers/mongo/server-mongo.js [hunk 8]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
+  let params = req.params;
+
+  let q_id = Number(params.question_id)
+
+  db.reportQuestion(q_id)
+  .then(() => {
+    res.status(204).send()
+  })
+  .catch(err => {
+    console.log('error marking question reported', err)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:123]: Maybe add res.sendStatus?

File: servers/mongo/server-mongo.js [hunk 9]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
 
-app.put('qa/answers/[answerid]/helpful')
+  let a_id = Number(params.answer_id);
+
+  db.markAnswerHelpful(a_id)
+  .then(() => {
+    res.status(204).send()
+  })
+  .catch(err => {
+    console.log('error marking answer helpful', err)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:137]: Maybe add res.sendStatus?

File: servers/mongo/server-mongo.js [hunk 10]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
+  let params = req.params;
+
+  let a_id = Number(params.answer_id);
+
+  db.reportAnswer(a_id)
+  .then(() => {
+    res.status(204).send()
+  })
+  .catch(err => {
+    console.log('error marking answer reported', err)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:151]: Maybe add res.sendStatus?

File: servers/mongo/server-mongo.js [hunk 11]
```diff
@@ -11,24 +11,148 @@ app.use(express.urlencoded({ extended: true }));
...
+  let email = query.email;
+  let photos = query.photos || []
 
+  if(typeof query.photos === 'string') {
+    try {
+      photos = query.photos.split('').filter(l => {
+        return (l !== '[' && l !== ']')
+      }).join('').split(',')
+
+      console.log('photos', photos)
```
Comment (jenniferjinyoungyang) [inline servers/mongo/server-mongo.js:80]: Do you still need this console log?

File: package.json [hunk 12]
```diff
@@ -4,8 +4,8 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "test": "jest",
-    "start": "node dbs/mongo/index-mongo.js"
+    "test": "jest --coverage",
+    "start": "node dbs/mongo/index.js"
```
Comment (jenniferjinyoungyang) [inline package.json:8]: Maybe use nodemon? And dbs/mongo/index.js is just a script with a bunch of routes. Is this the one you want to run with npm start?

File: dbs/mongo/queries/getAnswerData.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
+const {Answers, Answer_Photos} = require('../models.js');
+const {formatAnswerResponse} = require('../controllers/answerResponse.js')
+
+async function getAnswerData(q_id, page, count) {
+
+  let answers = await Answers.find({question_id: q_id, reported:0}, {_id: 0}).limit(count);
+
+  let answerPhotos = await Promise.all(answers.map(async (answer) => {
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/getAnswerData.js:18]: Is it necessary to use Promise.all when you are using async - await? And where those promises get resolved / rejected?

File: dbs/mongo/queries/getQuestionData.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+  let questions = await Questions.find({product_id:p_id, reported: 0}, {_id: 0}).limit(count);
+
+  let answers = await Promise.all(questions.map(async (question) => {
+    let answer = await Answers.find({question_id: question.question_id, reported: 0}, {_id: 0}).limit(count);
+    return answer
+  }))
+
+  let answerPhotos = await Promise.all(answers.map(async (answerGroup) => {
+    let photos = await Promise.all(answerGroup.map(async (answer) => {
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/getQuestionData.js:24]: Is it necessary to use Promise.all when you are using async - await? And where those promises get resolved / rejected?

File: dbs/mongo/queries/getQuestionData.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
+const {Questions, Answers, Answer_Photos} = require('../models.js');
+const {formatResponse} = require('../controllers/questionResponse.js');
+
+async function getQuestionData(p_id, page, count) {
+
+  let questions = await Questions.find({product_id:p_id, reported: 0}, {_id: 0}).limit(count);
+
+  let answers = await Promise.all(questions.map(async (question) => {
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/getQuestionData.js:18]: Is it necessary to use Promise.all when you are using async - await? And where those promises get resolved / rejected?

File: dbs/mongo/queries/addQuestion.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+const {Questions} = require('../models.js');
+
+async function addQuestion(p_id, name, email, body){
+
+  //let findHighestQ_ID = await Questions.find().sort({question_id: -1}).limit(1);
+
+ // let q_id =  findHighestQ_ID[0].question_id + 1;
+
+  let doc = new Questions({
+    question_id: 1,
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/addQuestion.js:10]: question_id is always 1?

File: dbs/mongo/queries/addQuestion.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    asker_name: name,
+    asker_email: email,
+    reported: 0,
+    question_helpfulness: 0
+  })
+
+  //console.log('Document saved',document)
+
+  await doc.save()
+  return `saved document`
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/addQuestion.js:23]: Do you need to return this string?

File: dbs/mongo/queries/addAnswer.js [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+const {addAnswerPhoto} = require('./addPhotos.js')
+
+async function addAnswer(q_id, body, name, email, photos){
+  let findHighestA_ID = await Answers.find().sort({answer_id: -1}).limit(1);
+
+  let a_id = findHighestA_ID[0].answer_id + 1;
+
+  if (photos.length > 0) {
+    await Promise.all(photos.map(async (photo) => {
+      await addAnswerPhoto(a_id, photo)
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/addAnswer.js:11]: Is it necessary to use Promise.all when you are using async - await? And where those promises get resolved / rejected?

File: dbs/mongo/index.js [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+const getAllQuestions = (p_id, page = 1, count = 5) => {
+  return getQuestionData(p_id, page, count);
+}
+
+const getAnswers = (q_id, page = 1, count = 5) => {
+  return getAnswerData(q_id, page, count)
+}
+
+const postQuestion = (p_id, name, email, body) => {
+  return addQuestion(p_id, name, email, body)
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/index.js:26]: Do you need to return anything?

File: dbs/mongo/index.js [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+const postAnswer = (q_id, body, name, email, photos) => {
+  return addAnswer(q_id, body, name, email, photos)
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/index.js:30]: Do you need to return anything?

File: dbs/mongo/index.js [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+const markQuestionHelpful = (q_id) => {
+  try{
+    return markQuestion(q_id, 'helpful')
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/index.js:35]: Do you need to return anything?

File: dbs/mongo/index.js [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+  } catch(e) {
+    console.log(e)
+  }
+
+}
+
+const reportQuestion = (q_id) => {
+  return markQuestion(q_id, 'reported')
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/index.js:43]: Do you need to return anything?

File: dbs/mongo/index.js [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+const markAnswerHelpful = (a_id) => {
+  return markAnswer(a_id, 'helpful')
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/index.js:47]: Do you need to return anything?

File: dbs/mongo/index.js [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+const reportAnswer = (a_id) => {
+  return markAnswer(a_id, 'reported')
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/index.js:51]: Do you need to return anything?

File: dbs/mongo/queries/markQuestion.js [hunk 25]
```diff
@@ -0,0 +1,27 @@
+const {Questions} = require('../models.js');
+
+async function markQuestion(q_id, status){
+  if (!q_id && typeof q_id !== 'number') {
+    throw 'Invalide question id'
+  }
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/markQuestion.js:6]: Do you need same validation for markAnswer function?

File: dbs/mongo/queries/markQuestion.js [hunk 26]
```diff
@@ -0,0 +1,27 @@
...
+
+  if (status === 'helpful') {
+   await Questions.findOneAndUpdate({question_id: q_id}, {$inc : {question_helpfulness : 1}})
+   .then(res => {
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/markQuestion.js:10]: await and then block? Is findOneAndUpdate a promise? If so, do you need await?

File: dbs/mongo/queries/markQuestion.js [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+     if (!res) {
+       throw new Error('Question does not exist')
+     }
+   })
+  } else if (status === 'reported') {
+    await Questions.findOneAndUpdate({question_id: q_id}, {$inc: {reported: 1}})
+    .then(res => {
```
Comment (jenniferjinyoungyang) [inline dbs/mongo/queries/markQuestion.js:17]: await and then block? Is findOneAndUpdate a promise? If so, do you need await?

File: start.js [hunk 28]
```diff
@@ -0,0 +1,8 @@
+const app = require('./servers/mongo/server-mongo.js');
+
+const mongoose = require('mongoose')
+mongoose.connect('mongodb://localhost/qa', {useNewUrlParser: true, useUnifiedTopology: true, useFindAndModify: false, autoIndex: false});
+
+const PORT = 3000;
+
+app.listen(PORT, () => {console.log(`Listening at ${PORT}`)})
```
Comment (jenniferjinyoungyang) [inline start.js:8]: Where does this script get used?

**Review by jenniferjinyoungyang — APPROVED** [review]
There were some semi colons missing and extra space inconsistent. Are you using a linter?

File: testHelpers/hasAllKeys.js (hunk 1 above)
Comment (tbrinkman3) [reply testHelpers/hasAllKeys.js:7]: It might be?  I changed it to .includes instead.  Thank you!

File: servers/mongo/server-mongo.js (hunk 5 above)
Comment (tbrinkman3) [reply servers/mongo/server-mongo.js:86]: Ah good point, I will send a 400 response in the catch, thank you!

File: dbs/mongo/queries/addQuestion.js (hunk 17 above)
Comment (tbrinkman3) [reply dbs/mongo/queries/addQuestion.js:23]: No I do not. I changed it to return the saved document, thanks!

File: package.json (hunk 12 above)
Comment (tbrinkman3) [reply package.json:8]: I haven't been using the start script, I went ahead and got rid of it.  I did install nodemon but when I try and run it is says the command was not found so I didn't bother debugging it and just enter the command manually....

File: dbs/mongo/queries/getQuestionData.js (hunk 14 above)
Comment (tbrinkman3) [reply dbs/mongo/queries/getQuestionData.js:24]: I do not know if it is the right way, but without the promise all, I do not get the responses back, so I needed to make sure that all data was received before moving on.

File: dbs/mongo/queries/addQuestion.js (hunk 16 above)
Comment (tbrinkman3) [reply dbs/mongo/queries/addQuestion.js:10]: nope, that definitely needs to be changed, thank you!

File: dbs/mongo/queries/markQuestion.js (hunk 25 above)
Comment (tbrinkman3) [reply dbs/mongo/queries/markQuestion.js:6]: yes, added!

File: dbs/mongo/queries/markQuestion.js (hunk 26 above)
Comment (tbrinkman3) [reply dbs/mongo/queries/markQuestion.js:10]: You are right, not needed, just got async/await happy

File: start.js (hunk 28 above)
Comment (tbrinkman3) [reply start.js:8]: that is for my 'production' db and i have another one for my stress tests

**Comment by tbrinkman3** [comment]
Thanks for looking through this!  I appreciate you catching all of those issues!

**Pull request merged**